File: BottomBarJava/src/main/java/com/ooftf/bottombar/java/FragmentSwitchManager.java
package com.ooftf.bottombar.java;

import androidx.annotation.NonNull;
import androidx.fragment.app.Fragment;
import androidx.fragment.app.FragmentManager;
import androidx.fragment.app.FragmentTransaction;

import java.util.Set;


/**
 * Created by master on 2017/9/27 0027.
 * 要保证T的类型在内存回收之后再次创建其toString值不会改变
 */
public class FragmentSwitchManager<T> {
    private FragmentManager manager;
    private int containerViewId;
    private Set<T> tags;
    private OnFragmentCommitListener<T> OnFragmentCommitListener;
    private FragmentCreator<T> createFragment;

    public FragmentSwitchManager(@NonNull FragmentManager manager, @NonNull int containerViewId, @NonNull Set<T> tags, @NonNull FragmentCreator<T> creator) {
        this.manager = manager;
        this.createFragment = creator;
        this.containerViewId = containerViewId;
        this.tags = tags;
    }


    public void setOnFragmentCommitListener(OnFragmentCommitListener<T> onFragmentCommitListener) {
        this.OnFragmentCommitListener = onFragmentCommitListener;
    }

    public void switchFragment(T tagId) {
        FragmentTransaction fragmentTransaction = manager.beginTransaction();
        hideOther(fragmentTransaction, tagId);
        Fragment fragment = getFragment(fragmentTransaction, tagId);
        if (OnFragmentCommitListener != null) {
            OnFragmentCommitListener.onFragmentCommit(tagId, fragment);
        }
        fragmentTransaction.show(fragment).commitAllowingStateLoss();
        manager.executePendingTransactions();
    }

    private void hideOther(FragmentTransaction fragmentTransaction, T tagId) {
        for (T tag : tags) {
            if (tag.equals(tagId)) {
                continue;
            }
            Fragment fragment = manager.findFragmentByTag(tag.toString());
            if (fragment != null) {
                if (!fragment.isAdded()) {
                    fragmentTransaction.add(containerViewId, fragment, tag.toString());
                }
                fragmentTransaction.hide(fragment);
            }

        }
    }

    private Fragment getFragment(FragmentTransaction fragmentTransaction, T tagId) {
        Fragment fragment = manager.findFragmentByTag(tagId.toString());
        if (null == fragment) {
            fragment = createFragment.create(tagId);
        }
        if (fragment != null && !fragment.isAdded()) {
            fragmentTransaction.add(containerViewId, fragment, tagId.toString());
        }
        return fragment;
    }
}

File: BottomBarJava/src/main/java/com/ooftf/bottombar/java/FragmentSwitchManager.kt
package com.ooftf.bottombar.java

import android.os.Handler
import android.os.Looper
import androidx.fragment.app.Fragment
import androidx.fragment.app.FragmentManager
import androidx.fragment.app.FragmentTransaction

/**
 * Created by master on 2017/9/27 0027.
 * 要保证T的类型在内存回收之后再次创建其toString值不会改变
 */
class FragmentSwitchManager<T>(
    private val manager: FragmentManager,
    private val containerViewId: Int,
    private val tags: Set<T>,
    private val createFragment: FragmentCreator<T>
) {
    private val fragmentTemp = HashMap<T, Fragment>()
    private val handler by lazy {
        Handler(Looper.getMainLooper())
    }


    fun switchFragment(tagId: T) {
        val fragmentTransaction = manager.beginTransaction()
        hideOther(fragmentTransaction, tagId)
        var fragment = findAddFragment(tagId)
        if (null == fragment) {
            fragment = createFragment.create(tagId)
            fragmentTemp[tagId] = fragment
            fragmentTransaction.add(containerViewId, fragment, tagId.toString())
        }
        fragmentTransaction.show(fragment).commitAllowingStateLoss()
        handler.post {
            fragmentTemp.remove(tagId)
        }
    }

    private fun hideOther(fragmentTransaction: FragmentTransaction, tagId: T) {
        for (tag in tags) {
            if (tag == tagId) {
                continue
            }
            findAddFragment(tag)?.let {
                fragmentTransaction.hide(it)
            }
        }
    }

    private fun findAddFragment(tagId: T): Fragment? {
        var fragment = manager.findFragmentByTag(tagId.toString())
        if (null == fragment) {
            fragment = fragmentTemp[tagId]
        }
        return fragment
    }
}